### ISSUE TITLE
Ambiguous Map method with Minimal

### ISSUE THREAD

**Issue by BrennanConroy** [dotnet/aspnetcore#35323]
## Proposed API

Add a new Map overload to remove ambiguity caused by implicit conversion from `PathString`:

```diff
namespace Microsoft.AspNetCore.Builder
{
    public static class MapExtensions
    {
+        public static IApplicationBuilder Map(this IApplicationBuilder app, string pathMatch, Action<IApplicationBuilder> configuration)
```

## Original Issue
When using minimal there is ambiguity when using `app.Map("/", builder => {});`
between `IApplicationBuilder Map(this IApplicationBuilder app, PathString pathMatch, Action<IApplicationBuilder> configuration)`
https://github.com/dotnet/aspnetcore/blob/ff51fd7105a9003841215f1f3b0b8fc9e2998a67/src/Http/Http.Abstractions/src/Extensions/MapExtensions.cs#L23
and `public static MinimalActionEndpointConventionBuilder Map(this IEndpointRouteBuilder endpoints, string pattern, Delegate action)`
https://github.com/dotnet/aspnetcore/blob/ff51fd7105a9003841215f1f3b0b8fc9e2998a67/src/Http/Routing/src/Builder/MinimalActionEndpointRouteBuilderExtensions.cs#L123

![image](https://user-images.githubusercontent.com/7574801/129303112-b1201e00-4ae7-4f43-ab48-7b9bf9caea31.png)

**Comment by davidfowl** [comment]
We had @filipw hit this in his blog post https://www.strathweb.com/2021/08/beautiful-and-compact-web-apis-revisited-with-c-10-and-net-6-preview-7/. Seems like we should discuss the options and workarounds here:

Work arounds:

1. Cast to IApplicationBuilder to call the overload of Map.

```C#
((IApplicationBuilder)app).Map("/openid", builder =>
{
    builder.UseIdentityServer();
});

// OR

IApplicationBuilder appBuilder = app;

appBuilder.Map("/openid", builder =>
{
    builder.UseIdentityServer();
});
```

2. Use the `IApplicationBuilder.Map` overload with 3 arguments (as the blog post does).

```C#
app.Map("/openid", true, builder =>
{
    builder.UseIdentityServer();
});
```

Possible solution: Add a Map overload directly to the `WebApplication` type that will win over the extension methods to avoid this ambiguity. This is something we should at the very least document.

**Comment by davidfowl** [comment]
@cston Could this be an issue with overload resolution. I just noticed the conflict is between 2 different extension methods that target interfaces, both of which a concrete type implement. I tried to replicate this with a minimal repro but it doesn't fail (I wonder if it's because they are in the same project):

```C#

var app = new WebApp();

app.Map("/sub", builder =>
{
    
});


class WebApp : IAppBuilder, IRouteBuilder
{

}

public interface IAppBuilder
{

}

public interface IRouteBuilder
{

}

static class ThingExtensions
{
    public static IAppBuilder Map(this IAppBuilder app, string path, Action<IAppBuilder> callback) => app;
}

public static class RouteBuilderExtensions
{
    public static IRouteBuilder Map(this IRouteBuilder routes, string path, Delegate callback) => routes;
}
```

**Comment by davidfowl** [comment]
I filed an issue here https://github.com/dotnet/roslyn/issues/55691

**Comment by davidfowl** [comment]
It's fixed!

**Comment by filipw** [comment]
wow that was fast 🥁

**Comment by halter73** [comment]
I reopened https://github.com/dotnet/roslyn/issues/55691 because there's still some weirdness there if you specify `IApplicationBuilder` as the parameter type in the lambda, but I think the main scenario where the parameter type is unspecified or specified to be anything other than `IApplicationBuilder` is addressed, so I'll leave this closed.

**Comment by halter73** [comment]
We can probably improve this more by adding a `string` overload to the `IApplicationBuilder` method in addition to the existing `PathString` overload.

**Comment by halter73** [comment]
Proposed new overload to remove ambiguity caused by implicit conversion from `PathString`:

```diff
namespace Microsoft.AspNetCore.Builder
{
    public static class MapExtensions
    {
+        public static IApplicationBuilder Map(this IApplicationBuilder app, string pathMatch, Action<IApplicationBuilder> configuration)
```

**Comment by halter73** [comment]
We decided this isn't worth the minor risk given that dotnet/roslyn#55357 has been merged which fixes the main scenario without any new overloads.

**Comment by halter73** [comment]
Fixed by #35631 #36044
